bot/code_coverage_bot/taskcluster.py

```diff
@@ -33,75 +25,63 @@ def get_task(branch, revision, platform):
-        return task["taskId"]
-    else:
-        if task["code"] == "ResourceNotFound":
+        raise Exception(f"Unsupported platform: {platform}")
+
+    route = f"gecko.v2.{branch}.revision.{revision}.{product}.{platform_name}"
+    index = taskcluster_config.get_service("index")
+    try:
+        return index.findTask(route)["taskId"]
+    except taskcluster.exceptions.TaskclusterRestFailure as e:
+        if e.status_code == 404:
             return None
-        else:
-            raise TaskclusterException("Unknown TaskCluster index error.")
+        raise
 
 
 def get_task_details(task_id):
-    r = requests.get(queue_base + "task/{}".format(task_id))
-    r.raise_for_status()
-    return r.json()
+    queue = taskcluster_config.get_service("queue")
+    return queue.task(task_id)
 
 
 def get_task_status(task_id):
-    r = requests.get(queue_base + "task/{}/status".format(task_id))
-    r.raise_for_status()
-    return r.json()
+    queue = taskcluster_config.get_service("queue")
+    return queue.status(task_id)
 
 
 def get_task_artifacts(task_id):
-    r = requests.get(queue_base + "task/{}/artifacts".format(task_id))
-    r.raise_for_status()
-    return r.json()["artifacts"]
+    queue = taskcluster_config.get_service("queue")
+    return queue.listLatestArtifacts(task_id)["artifacts"]
 
 
 def get_tasks_in_group(group_id):
-    list_url = queue_base + "task-group/{}/list".format(group_id)
-
-    r = requests.get(list_url, params={"limit": 200})
-    r.raise_for_status()
-    reply = r.json()
-    tasks = reply["tasks"]
-    while "continuationToken" in reply:
-        r = requests.get(
-            list_url,
-            params={"limit": 200, "continuationToken": reply["continuationToken"]},
-        )
-        r.raise_for_status()
-        reply = r.json()
-        tasks += reply["tasks"]
-    return tasks
+    queue = taskcluster_config.get_service("queue")
+
+    token = None
+    while True:
+        query = {"limit": 200}
+        if token is not None:
+            query["continuationToken"] = token
+
+        response = queue.listTaskGroup(group_id, query=query)
+
+        for task in response["tasks"]:
+            yield task
+
+        token = response.get("continuationToken")
+        if token is None:
+            break
 
 
 def download_artifact(artifact_path, task_id, artifact_name):
     if os.path.exists(artifact_path):
         return artifact_path
 
-    def perform_download():
-        r = requests.get(
-            queue_base + "task/{}/artifacts/{}".format(task_id, artifact_name),
-            stream=True,
-        )
-
-        r.raise_for_status()
+    queue = taskcluster_config.get_service("queue")
 
+    def perform_download():
+        response = queue.getLatestArtifact(task_id, artifact_name)
         with open(artifact_path, "wb") as f:
-            r.raw.decode_content = True
-            shutil.copyfileobj(r.raw, f)
+            content = response["response"].content
+            f.write(content)
```

Comment:
This used to stream the response directly to the file, it should be better in terms of memory usage when we download many files.
Is there a way to implement that with the taskcluster lib?

Reply:
I restored it by building the artifact url using taskcluster lib.